### PR TITLE
Add a matmul test from int8, bf16

### PR DESCRIPTION
**Pull request by karupayun**
In this PR we are adding a matmul test from int8, bf16. I had a few issues in the test so
I refactored the file a bit.
    
- First I included two new params:
  - Dot_out_dtype: So users of the test class can specify the type used internally in the dot,
     and not the one set by default given the two types. There are several restrictions for
     these types anyway.
  - C_dtype: The return type of the matmul.
I included a few tests in the case of making a dot with two float16.

 - I had to modify test_matmul to use small integers when testing with two float16 since torch
    used float32 internally in this case and we were having precision issues when comparing
    the results with triton in the case that dot_out_dtype was float16.
 - I also needed to include torch.int8 in the possible datatypes.

Finally I tried to simplify a bit the logic of the matmul/test_matmul because after adding
these two parameters it was a bit hard to follow why we needed every part of the code, so
I included a type_preference_list for the allowed dot_out_dtype given the types of the operands
a and b.

### CODE REVIEW

File: python/test/unit/operators/test_matmul.py [hunk 1]
```diff
@@ -108,7 +117,7 @@ def kernel(Y, X, N, BLOCK_SIZE: tl.constexpr):
         ],
     ),
 )
-def test_op(BLOCK_M, BLOCK_N, BLOCK_K, SPLIT_K, NWARP, NSTAGE, M, N, K, AT, BT, ADTYPE, BDTYPE, ALLOW_TF32, F8_FASTACCUM):
+def test_op(BLOCK_M, BLOCK_N, BLOCK_K, SPLIT_K, NWARP, NSTAGE, M, N, K, AT, BT, ADTYPE, BDTYPE, ALLOW_TF32, F8_FASTACCUM, DOT_OUT_DTYPE, C_DTYPE):
```
Comment (gflegar) [inline python/test/unit/operators/test_matmul.py:120]: I would rather replace `DOT_OUT_DTYPE` with `ACCUMULATION_DTYPE` or something like that. Otherwise it's confusing, since `dot` often refers to the entire matmul as well.

File: python/triton/ops/matmul.py [hunk 2]
```diff
@@ -144,7 +145,7 @@ class _matmul(torch.autograd.Function):
     _locks = {}
 
     @staticmethod
-    def _call(a, b, dot_out_dtype, allow_tf32, fp8_fast_accum):
+    def _call(a, b, dot_out_dtype, allow_tf32, fp8_fast_accum, c_dtype):
```
Comment (gflegar) [inline python/triton/ops/matmul.py:148]: rename `dot_out_type` to `accumulation_type`

File: python/triton/ops/matmul.py [hunk 3]
```diff
@@ -155,33 +156,41 @@ def _call(a, b, dot_out_dtype, allow_tf32, fp8_fast_accum):
         assert a.shape[1] == b.shape[0], "incompatible dimensions"
         M, K = a.shape
         _, N = b.shape
+
+        # common type between a and b
+        ab_dtype = get_higher_dtype(a.dtype if a.dtype not in [tl.float8e4nv, tl.float8e4b15, tl.float8e5] else torch.float16,
```
Comment (gflegar) [inline python/triton/ops/matmul.py:161]: nit: does something like `"float8" in str(a.dtype)` work instead of checking for this list?

File: python/triton/ops/matmul.py [hunk 4]
```diff
@@ -155,33 +156,41 @@ def _call(a, b, dot_out_dtype, allow_tf32, fp8_fast_accum):
...
         c = torch.empty((M, N), device=device, dtype=c_dtype)
+
+        torch_tl_type_converter = {torch.float16: tl.float16, torch.bfloat16: tl.bfloat16, torch.float32: tl.float32, torch.int8: tl.int8, torch.int32: tl.int32}
+
+        # Allowed types for dot_out_type given the types of a and b.
+        type_preference_list = {}
+        type_preference_list[torch.float16] = [torch.float32, torch.float16]
+        type_preference_list[torch.bfloat16] = [torch.float32, torch.bfloat16]
+        type_preference_list[torch.float32] = [torch.float32]
+        type_preference_list[torch.int8] = [torch.int32]
```
Comment (gflegar) [inline python/triton/ops/matmul.py:177]: Use dictionary constructor (and tuples) instead, with a bit clearer naming:

```python
supported_accumulator_dtypes = {
  torch.float16: (torch.float32, torch.float16),
  ...
}
```

File: python/triton/ops/matmul.py [hunk 5]
```diff
@@ -155,33 +156,41 @@ def _call(a, b, dot_out_dtype, allow_tf32, fp8_fast_accum):
...
-        elif a.dtype in [torch.int8] or b.dtype in [torch.int8]:
-            c_dtype = torch.int32
-        else:
-            c_dtype = get_higher_dtype(a.dtype, b.dtype)
+        if (c_dtype is None):
+            c_dtype = ab_dtype
+
         c = torch.empty((M, N), device=device, dtype=c_dtype)
+
+        torch_tl_type_converter = {torch.float16: tl.float16, torch.bfloat16: tl.bfloat16, torch.float32: tl.float32, torch.int8: tl.int8, torch.int32: tl.int32}
```
Comment (gflegar) [inline python/triton/ops/matmul.py:170]: Would something like this work without spelling out the table manually:

```python
def to_tl_type(ty):
  return getattr(tl, str(ty).split(".")[-1])
```

File: python/triton/ops/matmul.py (hunk 3 above)
Comment (gflegar) [reply python/triton/ops/matmul.py:161]: Also, would it make sense to make these `if`s a part of the logic within `get_higher_dtype`?

File: python/triton/ops/matmul.py [hunk 6]
```diff
@@ -155,33 +156,41 @@ def _call(a, b, dot_out_dtype, allow_tf32, fp8_fast_accum):
...
+        type_preference_list[torch.float16] = [torch.float32, torch.float16]
+        type_preference_list[torch.bfloat16] = [torch.float32, torch.bfloat16]
+        type_preference_list[torch.float32] = [torch.float32]
+        type_preference_list[torch.int8] = [torch.int32]
         if dot_out_dtype is None:
-            if c_dtype in [torch.float16, torch.float32, torch.bfloat16]:
-                dot_out_dtype = tl.float32
-            else:
-                dot_out_dtype = tl.int32
+            dot_out_dtype = type_preference_list[c_dtype][0]
```
Comment (gflegar) [inline python/triton/ops/matmul.py:179]: `type_preference_list[ab_dtype][0]`? Why `c_dtype`?

File: python/triton/ops/matmul.py [hunk 7]
```diff
@@ -155,33 +156,41 @@ def _call(a, b, dot_out_dtype, allow_tf32, fp8_fast_accum):
...
-                dot_out_dtype = tl.int32
-        ab_dtype = True
+            assert dot_out_dtype in type_preference_list[a.dtype], "dot_out_dtype not compatible with the type of a"
+            assert dot_out_dtype in type_preference_list[b.dtype], "dot_out_dtype not compatible with the type of b"
+
+        dot_out_dtype = torch_tl_type_converter[dot_out_dtype]
+        ab_dtype = torch_tl_type_converter[ab_dtype]
+        c_dtype = torch_tl_type_converter[c_dtype]
+
+        # In this cases, there is no need to manually convert types to ab_dtype.
```
Comment (gflegar) [inline python/triton/ops/matmul.py:189]: Clarify why we do not need to convert instead of just saying what the code does, e.g.:

```python
# Tensor cores support input with mixed float8 types.
```

File: python/triton/ops/matmul.py [hunk 8]
```diff
@@ -155,33 +156,41 @@ def _call(a, b, dot_out_dtype, allow_tf32, fp8_fast_accum):
...
         if a.dtype in [tl.float8e4nv, tl.float8e5] and b.dtype in [tl.float8e4nv, tl.float8e5]:
-            ab_dtype = False
+            ab_dtype = None
         if a.dtype in [torch.int8] and b.dtype in [torch.int8]:
```
Comment (gflegar) [inline python/triton/ops/matmul.py:192]: Why do we need this if? If they're both already `torch.int8`, then `ab_dtype` should already be `torch.int8`, and the conversion is a no-op.

File: python/triton/ops/matmul.py [hunk 9]
```diff
@@ -84,7 +84,8 @@ def _kernel(A, B, C, M, N, K,
             allow_tf32: tl.constexpr,
             fp8_fast_accum: tl.constexpr,
             BLOCK_M: tl.constexpr, BLOCK_N: tl.constexpr, BLOCK_K: tl.constexpr,
-            GROUP_M: tl.constexpr, SPLIT_K: tl.constexpr, EVEN_K: tl.constexpr, AB_DTYPE: tl.constexpr
+            GROUP_M: tl.constexpr, SPLIT_K: tl.constexpr, EVEN_K: tl.constexpr,
+            AB_DTYPE: tl.constexpr, C_DTYPE: tl.constexpr
```
Comment (gflegar) [inline python/triton/ops/matmul.py:88]: No need to pass `C_DTYPE`, since we already have it through `C.dtype.element_ty`.

File: python/test/unit/operators/test_matmul.py (hunk 1 above)
Comment (gflegar) [reply python/test/unit/operators/test_matmul.py:120]: Also, `C_DTYPE` is a bit confusing as well, since it sort of assumes that we'll call the output of our `tt.ops.dot()` call "C". Maybe just call it `OUTPUT_DTYPE`.

**Review by gflegar — CHANGES_REQUESTED** [review]
Thanks! Definitely much simpler and more versatile than it was before. But I think it can be even simpler and clearer - left some comments to do that.

File: python/triton/ops/matmul.py (hunk 2 above)
Comment (karupayun) [reply python/triton/ops/matmul.py:148]: I will use acc_type

File: python/test/unit/operators/test_matmul.py [hunk 10]
```diff
@@ -137,22 +146,34 @@ def maybe_upcast(x, dtype, is_float8):
...
+                return torch.randint(-2, 2, (m, n), device="cuda", dtype=torch.int8)
             return torch.randint(-128, 127, (m, n), device="cuda", dtype=torch.int8)
-        dtype = {"float16": torch.float16, "bfloat16": torch.bfloat16, "float32": torch.float32}[dtype]
-        exponents = torch.randint(-10, 0, size=(m, n))
-        ret = (2. ** exponents).to(dtype).to("cuda")
+        if only_small_integers:
+            exponents = torch.randint(0, 3, size=(m, n))
+        else:
+            exponents = torch.randint(-10, 0, size=(m, n))
+        ret = (2. ** exponents).to(string_to_torch_type[dtype]).to("cuda")
```
[the remark this answers is not in the document]
Comment (karupayun) [reply python/test/unit/operators/test_matmul.py:165]: thanks! :)

File: python/test/unit/operators/test_matmul.py [hunk 11]
```diff
@@ -137,22 +146,34 @@ def maybe_upcast(x, dtype, is_float8):
...
             sign = torch.randint(2, size=(m, n), device="cuda", dtype=torch.int8) * 128
             val = torch.randint(2**3 - 1, size=(m, n), device="cuda", dtype=torch.int8) << 7 - ewidth
             return sign | val
         if dtype == "int8":
+            if only_small_integers:
+                return torch.randint(-2, 2, (m, n), device="cuda", dtype=torch.int8)
             return torch.randint(-128, 127, (m, n), device="cuda", dtype=torch.int8)
-        dtype = {"float16": torch.float16, "bfloat16": torch.bfloat16, "float32": torch.float32}[dtype]
-        exponents = torch.randint(-10, 0, size=(m, n))
-        ret = (2. ** exponents).to(dtype).to("cuda")
```
[the remark this answers is not in the document]
Comment (karupayun) [reply python/test/unit/operators/test_matmul.py:150]: They are not integers. Exponents go from -10 to 0, so from 1/1024 to 1.

File: python/test/unit/operators/test_matmul.py [hunk 12]
```diff
@@ -137,22 +146,34 @@ def maybe_upcast(x, dtype, is_float8):
...
+        else:
+            exponents = torch.randint(-10, 0, size=(m, n))
+        ret = (2. ** exponents).to(string_to_torch_type[dtype]).to("cuda")
         return ret
 
     # allocate/transpose inputs
-    a = init_input(M, K, ADTYPE)
-    b = init_input(K, N, BDTYPE)
+    use_small_integers = False
+    # We can't force torch to use float16 for internal computations (it will use float32), and
```
[the remark this answers is not in the document]
Comment (karupayun) [reply python/test/unit/operators/test_matmul.py:170]: I think that currently it IS the issue. Epsilon from float16 is around ~1e-3, way larger than the maximum expected error of 1e-5 so there is no chance that they will match using non-integers. So, it's never going to work. 

But I agree about that even using float16 both sides we could have potentially the issue because floating point arithmetic and associativity. I will rephrase the comment.

File: python/test/unit/operators/test_matmul.py [hunk 13]
```diff
@@ -137,22 +146,34 @@ def maybe_upcast(x, dtype, is_float8):
...
+    # therefore we need to use small integers if we want the comparation to be precise enough
+    # in case of a and b being float16.
+    if (ADTYPE == "float16" and BDTYPE == "float16"):
+        use_small_integers = True
```
[the remark this answers is not in the document]
Comment (karupayun) [reply python/test/unit/operators/test_matmul.py:174]: I did these two cases following the discussion in https://chat.google.com/room/AAAAUv6LytY/2-1nSny8Jfc. I could stay with always small integers but we lose the option of asking the test to be more precise.

I also could pass the ACC_DTYPE to the init_input constructor and use small integers in the case that it's not float32.

File: python/test/unit/operators/test_matmul.py [hunk 14]
```diff
@@ -164,19 +185,15 @@ def init_input(m, n, dtype):
     th_b = maybe_upcast(b, BDTYPE, b_fp8)
     if BT and b_fp8:
```
[the remark this answers is not in the document]
Comment (karupayun) [reply python/test/unit/operators/test_matmul.py:186]: Done. 

2 things. 
* I need the type of the upcast, so I still have the variable `th_a` and `th_b`.
* I do not understand why we are doing `th.view(th.shape[::-1]).T`. 

We can discuss about that, for now I left a TODO that I'm planning to erase after talking offline.

File: python/test/unit/operators/test_matmul.py [hunk 15]
```diff
@@ -164,19 +185,15 @@ def init_input(m, n, dtype):
...
         th_b = th_b.view(th_b.shape[::-1]).T
-    if th_a.is_floating_point():
-        ab_dtype = th_a.dtype if th_a.element_size() > th_b.element_size() else th_b.dtype
-    else:
-        ab_dtype = torch.float32
-    th_c = torch.matmul(th_a.to(ab_dtype), th_b.to(ab_dtype))
-    if ADTYPE == "int8" or BDTYPE == "int8":
-        th_c = th_c.to(torch.int8)
+    dot_dtype = string_to_torch_type[DOT_OUT_DTYPE] if DOT_OUT_DTYPE else triton.ops.get_higher_dtype(th_a.dtype, th_b.dtype)
+    c_dtype = string_to_torch_type[C_DTYPE] if C_DTYPE else dot_dtype
```
[the remark this answers is not in the document]
Comment (karupayun) [reply python/test/unit/operators/test_matmul.py:189]: done

File: python/triton/ops/matmul.py (hunk 6 above)
Comment (karupayun) [reply python/triton/ops/matmul.py:179]: Nice catch!

File: python/triton/ops/matmul.py (hunk 7 above)
Comment (karupayun) [reply python/triton/ops/matmul.py:189]: good suggestion!

File: python/triton/ops/matmul.py (hunk 8 above)
Comment (karupayun) [reply python/triton/ops/matmul.py:192]: Yes, you are completely right.

File: python/test/unit/operators/test_matmul.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (karupayun) [reply python/test/unit/operators/test_matmul.py]: I moved this logic to the generation of values. Let me know what you think. I'm open to use small integers everywhere.

[the remark this answers is not in the document]
Comment (gflegar) [reply python/test/unit/operators/test_matmul.py]: I still think we might not need to change this at all, as long as we randomly generate our values around 0 (by also randomly generating the sign), but that's the other comment thread.

File: python/test/unit/operators/test_matmul.py [hunk 16]
```diff
@@ -132,51 +124,69 @@ def test_op(BLOCK_M, BLOCK_N, BLOCK_K, SPLIT_K, NWARP, NSTAGE, M, N, K, AT, BT,
...
+        grid = lambda META: (triton.cdiv(x.numel(), META['BLOCK_SIZE']),)
+        dtype = getattr(tl, dtype)
+        kernel[grid](ret, triton.reinterpret(x, dtype), ret.numel(), BLOCK_SIZE=1024)
+        return ret
+
+    def upcast_if_fp8(x, dtype, is_transpose):
+        if is_fp8(dtype):
+            th = f8_to_f16(x, dtype)
+            if (is_transpose):
+                # TODO(karupayun): I do not understand why we are doing this.
```
[the remark this answers is not in the document]
Comment (karupayun) [reply python/test/unit/operators/test_matmul.py:150]: If I comment the if + view (line 149-151) and I modify the line as you proposed, the test fails.

File: python/test/unit/operators/test_matmul.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (karupayun) [reply python/test/unit/operators/test_matmul.py:150]: Ah, you are right! It's the same than with integers. Thanks for pointing me this out, I could use -3 to 0 for small cases (instead of 0 to 3) :)

However, I tried to generate randomly the sign and we move from 4 failing tests to 14 failing ones :(.

The new failing tests are ones with `bfloat16`, but also smaller cases (32x32) that before only failed with output_type `float32` but now also fails with `float16`:
```FAILED test/unit/operators/test_matmul.py::test_op[32-32-32-1-1-2-None-None-None-False-False-float16-float16-True-True-float16-None] - AssertionError: Tensor-likes are not close!```

 If I force the sign to be always negative results also in 4 failing tests, so no idea what could be happening. But also I really doubt that this will help since the positive and the negative numbers shouldn't cancel each other in this random generation.

[the remark this answers is not in the document]
Comment (karupayun) [reply python/test/unit/operators/test_matmul.py:150]: I would propose to have float16 case with -3 to 0, what do you think?

[the remark this answers is not in the document]
Comment (karupayun) [reply python/test/unit/operators/test_matmul.py:150]: Offline, we decided to use positives from -4 to 0, since we shouldn't be testing precision issues in this test

File: python/test/unit/operators/test_matmul.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (karupayun) [reply python/test/unit/operators/test_matmul.py:150]: This is fixed by modifying line 140 to `        ret = torch.empty_strided(x.shape, x.stride(), dtype=torch.float16, device=x.device)
`

File: python/test/unit/operators/test_matmul.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (karupayun) [reply python/test/unit/operators/test_matmul.py]: Not, they fail, and we need th_a and th_b types, so I will disregard this last suggestion since you mentioned as a nit :)

**Comment by karupayun** [comment]
This PR was divided between https://github.com/openai/triton/pull/2768, https://github.com/openai/triton/pull/2769 and https://github.com/openai/triton/pull/2760. All of them are already merged.
